napari/_vispy/layers/labels.py

```diff
@@ -57,21 +60,34 @@
 _UNSET = object()
 
 
-auto_lookup_shader = """
+auto_lookup_shader_uint8 = """
 uniform sampler2D texture2D_values;
 
 vec4 sample_label_color(float t) {
-    // VisPy automatically scales uint8 and uint16 to [0, 1].
-    // this line fixes returns values to their original range.
-    t = t * $scale;
+    if (($use_selection) && ($selection != int(t * 255))) {
+        return vec4(0);
+    }
+    return texture2D(
+        texture2D_values,
+        vec2(0.0, t)
+    );
+}
+"""
 
-    if (($use_selection) && ($selection != t)) {
+auto_lookup_shader_uint16 = """
+uniform sampler2D texture2D_values;
+
+vec4 sample_label_color(float t) {
+    // uint 16
+    t = t * 65535;
+    if (($use_selection) && ($selection != int(t))) {
         return vec4(0);
     }
-    t = mod(t, $color_map_size);
+    float v = mod(t, 256);
+    float v2 = (t- v) / 256;
```

Comment:
```suggestion
    float v2 = (t - v) / 256;
```

Introducing... BlackenGL ™️